### PR TITLE
Add Raku (Perl 6) filetype support (lexer and ctags parser)

### LINKED ISSUE #2456: Raku Syntax Highlighting
Scintilla 4.30 onward ships with a Raku lexer: https://sourceforge.net/p/scintilla/feature-requests/1328/

Would it be possible to add Raku support in Geany?

### LINKED ISSUE #3053: Raku comment #`{comment} not recognized properly
Fedora 35
geany-1.38-1.fc35.x86_64

Bug:  geany is turning the top open { bracket blue instead of the second open bracket  #`{ blue when the first close bracket } is highlighted.   

This bug really makes find missing closing brackets a pain in the neck!

```
sub GetWindows10Build() {
   my $SubName = &?ROUTINE.name;
   PrintRedErr( "$SubName not yet implemented\n" ); return;  
#`{
      multi line comments go here
  }

}
```

### LINKED ISSUE #3054: RFE: please add support for Perl6 (Raku)
Request for Enhancement: please add support for Perl6 (Raku)

### PR DESCRIPTION
**Pull request by techee**
Note: I have never used Perl 6 (and did my best to avoid Perl) so I largely don't know what I'm doing. There just are several open issues requesting this language and I believe Geany should support it. And even if this PR isn't perfect, it should be much easier for real Raku users to improve it once this PR is in (e.g. by providing better keywords or highlighting).

I based the keyword list on

https://github.com/Raku/vim-raku/blob/master/syntax/raku.vim

and also fixed one lexilla bug here

https://github.com/ScintillaOrg/lexilla/pull/76

Fixes #3054, fixes #3053, fixes #2456.

### CODE REVIEW

**Comment by techee** [comment]
@kugel- I added a patch updating HACKING with some changes related to adding a lexer and also mentioning updates to meson.build. Would you have a look if the changes look good to you?

**Comment by kugel-** [comment]
Speaking only for HACKING it seems alright

**Comment by techee** [comment]
I just repushed with all the keyword groups and keywords from Scintilla.

**Comment by elextr** [comment]
No Rakuist has tested in the time this has been here, maybe its not worth adding.

**Comment by rdipardo** [comment]
> No Rakuist has tested in the time this has been here

One way to promote user participation might be configuring the CI workflow to pack a binary tarball and save it with the [upload-artifact][0] action.

Not that Geany users need coddling, but source builds really are problematic when the `make install` step could clobber your system's editor, and hacking the `LD_LIBRARY_PATH` is not advisable (although I've found it [does work in a pinch][1]).


[0]: https://github.com/actions/upload-artifact
[1]: https://github.com/geany/geany/issues/3054#issuecomment-1237594886

**Comment by elextr** [comment]
> CI workflow to pack a binary tarball and save it with the [upload-artifact](https://github.com/actions/upload-artifact) action.

The problem with this is "for what system"? All the Linuxes have their own quirks, and most distros have their own patches for Geany to meet their requirements.  Somewhat ironically Windows and Macos are more consistent, although (IIUC) both have occasional forward/backward compatibility lapses and are somewhat harder to build for.

Making a source tarball is more likely portable, but its not really much less work for the user than using the git repo.

> but source builds really are problematic when the make install step could clobber your system's editor

On Linux its actually quite easy to build a fully isolated copy of Geany from git that can't clobber the system install (because its installed without privileges).  I did once try to add the instructions to `README` or `INSTALL` (can't remember which) but it was sabotaged by other devs who kept adding more options and insisting it should not be in tree etc etc, so it was abandoned as it ended up no simpler.

Assuming you have installed the required tools (eg on Debuntu `build-essential`, `git`, `libgtk-3-dev`, `autoconf` and all its recommends and suggests).  Many Geany users being developers will have at least some of these installed already.

In your home

```
mkdir git-geany; cd git-geany
git clone https://github.com/geany/geany.git
cd geany
./autogen.sh --prefix=/the/full/absolute/path/to/git-geany --disable-html-docs
make install
cd ../bin; ./geany -c ../config
```

Any time you cd to `git-geany/geany` you can follow (copy and paste) the command line instructions to checkout and test the PR on github and `make install`ing again.  Each PR is in its own branch so they don't interfere.  So you can test lots of PRs!!!!! :-)

When you are finished or something breaks your repository (damn force pushes) just delete the whole `git-geany` tree, I find its faster to re-clone something as small as Geany than it is to fix broken git repos.

> hacking the LD_LIBRARY_PATH is not advisable

correct.

**Comment by techee** [comment]
> No Rakuist has tested in the time this has been here, maybe its not worth adding.

Well, even though I'm not a big fan of anything perl-related, I think Geany should support Raku. We have 3 open issues related to this language so people seem to be interested in it and I think it's a relatively common language (compared to some rather esoteric languages that Geany supports) and shouldn't be omitted.

This patch does the "hard work" (especially for someone who knows only scripting languages and has no experience with Geany hacking) and even if e.g. the keyword selection isn't perfect, once the support is in, any subsequent improvement should be much easier for Raku users.

**Comment by eht16** [comment]
Agree with @techee and I could give it a bit of testing, though this might takes a few weeks.
But my Perl is also as low as I managed to keep it.

**Comment by techee** [comment]
So I guess this PR will have to wait for the next release, right? I'm a bit sad about that because Perl 6 was requested several times and this PR was waiting for a year and a half for merging but I guess it's too late now...

Should I extract the changes from HACKING to a separate PR? It contains the most up-to-date info about how to add ctags parsers and Scintilla lexers.

**Comment by b4n** [comment]
Do we want this in 2.0?

**Comment by b4n** [comment]
@techee yeah I'd think it's wiser to wait, but probably get it in soon after.  And yeah, HACKING changes are probably useful nonetheless.

**Comment by elextr** [comment]
@eht16 if you can test this thats great, its the main thing missing which delayed it until it was too late for release.  Since @techee has guaranteed he will support all these new languages lets take advantage of it :smile:

**Comment by eht16** [comment]
> Agree with @techee and I could give it a bit of testing, though this might takes a few weeks. 

s/few weeks/few years/ but we are all patient :).

I gave it some testing and it works fine. I want to repeat I'm really bad at Perl and saw Raku the first time now. Btw, I did not fell in love with it :D.

The perl6.c parser is taken from uctags, I assume. If so, we probably should to rename it as it was done in https://github.com/universal-ctags/ctags/pull/3777.

Geany detects https://github.com/Raku/examples/blob/master/categories/games/blackjack.p6 as Perl even the file extension tells Raku. This is because of the shebang in the file. The following change should solve it by first checking for "perl6" and only then "perl".

```diff
@@ -604,6 +605,7 @@ static GeanyFiletype *find_shebang(const gchar *utf8_filename, const gchar *line
                        { "sh",         GEANY_FILETYPES_SH },
                        { "bash",       GEANY_FILETYPES_SH },
                        { "dash",       GEANY_FILETYPES_SH },
+                       { "perl6",      GEANY_FILETYPES_RAKU },
                        { "perl",       GEANY_FILETYPES_PERL },
                        { "python",     GEANY_FILETYPES_PYTHON },
                        { "php",        GEANY_FILETYPES_PHP },
```

**Comment by techee** [comment]
> I gave it some testing and it works fine. I want to repeat I'm really bad at Perl and saw Raku the first time now. Btw, I did not fell in love with it :D.

Good, I was worried we'd have to rewrite Geany in it :-)

> The perl6.c parser is taken from uctags, I assume. If so, we probably should to rename it as it was done in https://github.com/universal-ctags/ctags/pull/3777.

I'm aware of that, I was just thinking it could be done during the next "ctags sync" if it's not a problem.

> Geany detects https://github.com/Raku/examples/blob/master/categories/games/blackjack.p6 as Perl even the file extension tells Raku. This is because of the shebang in the file.

I totally missed that one, thanks! I've just made a fixup commit with the proposed fix.

Is anything else missing? If there are no other problems, I'd like to merge this PR so we are finally done with it.

File: src/filetypes.c
Comment (b4n) [inline src/filetypes.c:608]: Given the run command is `raku "%f"`, shouldn't `raku` be included in this list as well?

**Review by b4n — COMMENTED** [review]
I have no clue about Rakut, but looks pretty good (as expected), minor comments/questions inline.

File: data/filedefs/filetypes.raku
Comment (techee) [reply data/filedefs/filetypes.raku:50]: Yeah, no clue either. After not finding it in shared-mime-info I just googled for "text/x-perl6" and "text/x-raku" and the former returns about 300 results while the latter returns just 5. But it could just be that the various sources haven't been renamed after the language name change.

File: src/highlightingmappings.h
Comment (techee) [reply src/highlightingmappings.h:1435]: If I understand it correctly, TRUE indicates the group into which dynamically obtained "keywords" from ctags are mapped. For C it's the various types like ScintillaObject or GeanyDocument, etc. which are then highlighted in the editor with the given style. I haven't seen it used by any scripting language though and there's a danger that the mapping of the given language, in our case
```C
static TMParserMapEntry map_RAKU[] = {
	{'c', tm_tag_class_t},      // class
	{'g', tm_tag_struct_t},     // grammar
	{'m', tm_tag_method_t},     // method
	{'o', tm_tag_namespace_t},  // module
	{'p', tm_tag_package_t},    // package
	{'r', tm_tag_class_t},      // role
	{'u', tm_tag_variable_t},   // rule
	{'b', tm_tag_method_t},     // submethod
	{'s', tm_tag_function_t},   // subroutine
	{'t', tm_tag_variable_t},   // token
};
```

doesn't correctly match the hard-coded "typedef" types from the tag manager:

```C
static TMTagType TM_GLOBAL_TYPE_MASK =
	tm_tag_class_t | tm_tag_enum_t | tm_tag_interface_t |
	tm_tag_struct_t | tm_tag_typedef_t | tm_tag_union_t | tm_tag_namespace_t;
```

(like mapping `grammar` to `struct` in the raku case - it probably is alright, I guess it's kind of type, but there isn't a guaranteed 1:1 correspondence)

Also, these types don't appear in function prototypes or variable declarations as these are dynamically-typed languages so I think it's not really that useful to have them colorized.

File: src/filetypes.c
Comment (techee) [reply src/filetypes.c:609]: Yeah, definitely, I've just added that.

**Comment by eht16** [comment]
>     I gave it some testing and it works fine. I want to repeat I'm
>     really bad at Perl and saw Raku the first time now. Btw, I did not
>     fell in love with it :D.
> 
> Good, I was worried we'd have to rewrite Geany in it :-)

Maybe I'll change my mind in two weeks, what about Reany?

:)

>     The perl6.c parser is taken from uctags, I assume. If so, we
>     probably should to rename it as it was done in
>     universal-ctags/ctags#3777
>     <https://github.com/universal-ctags/ctags/pull/3777>.
> 
> I'm aware of that, I was just thinking it could be done during the next 
> "ctags sync" if it's not a problem.

Alright, sounds good.

> Is anything else missing? If there are no other problems, I'd like to 
> merge this PR so we are finally done with it.

I'm fine with it and when @b4n is also, let's get it in.

File: src/highlightingmappings.h
Comment (b4n) [reply src/highlightingmappings.h:1435]: You convinced me that we can at least wait for Rakuers to request/do it if they see fit.

FWIW I would think it fairly nice to have e.g. user-defined Python classes highlighted just like types in C -- not a must, but I wouldn't find this out of place -- and could make sense with annotations.  Anyway, just saying that us not currently doing it for any scripting language is not necessarily a reason to keep it that way :)

File: data/filedefs/filetypes.raku
Comment (b4n) [reply data/filedefs/filetypes.raku:50]: Fair enough, and I guess it a MIME type pops up it'll support this alias anyway -- and we can updated it later on if something appears out of the fog at some point.

**Review by b4n — APPROVED** [review]
I still have no clue about Raku, but LGTM and works as far as super-basic testing goes.

**Comment by b4n** [comment]
> Maybe I'll change my mind in two weeks, what about Reany?

Sounds like Geany written in R, doesn't it? :)

**Comment by elextr** [comment]
> > Maybe I'll change my mind in two weeks, what about Reany?
> Sounds like Geany written in R, doesn't it? :)

Surely he means Racket, its [lisp](https://xkcd.com/297/) (well Scheme, but similar) and it has its own portable native GUI library built in, no more GTK on windows issues .. hmm I just might convince myself at this rate :-)

**Comment by techee** [comment]
Meh, I forgot how github handles fixup commits and kind of assumed they get auto-squashed when merging a PR (which didn't happen). So there's a slight mess as the fixup commits were preserved. Sorry for that.

**Comment by elextr** [comment]
The dropdown to the right of the big green button has a squash and merge option.

**Pull request merged**
